packages/engine/src/scene/systems/ShadowSystem.ts

```diff
@@ -96,7 +116,73 @@ export default async function ShadowSystem(world: World) {
     return null
   })
 
+  const shadowComponentQuery = defineQuery([DropShadowComponent])
+
+  const shadowOffset = new Vector3(0, 0.01, 0)
+  const shadowGeometry = new PlaneGeometry(1, 1, 1, 1)
+  const loader = new TextureLoader()
+  const texture = loader.load(
+    (process.env['VITE_FILE_SERVER'] ?? 'https://localhost:8642') + '/projects/default-project/public/drop-shadow.png'
+  )
+  const shadowMaterial = new MeshBasicMaterial({ map: texture, side: DoubleSide, transparent: true })
+  let dropShadows = new InstancedMesh(shadowGeometry, shadowMaterial, shadowComponentQuery().length)
+
+  let sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
+  console.log(sceneObjects)
+
+  const CreateDropShadows = () => {
+    const query = shadowComponentQuery()
+
+    if (dropShadows && query.length != dropShadows.count) {
+      world.scene.remove(dropShadows)
+      dropShadows = new InstancedMesh(shadowGeometry, shadowMaterial, shadowComponentQuery().length)
+      dropShadows.matrixAutoUpdate = false
+      dropShadows.layers.set(ObjectLayers.DropShadowCaster)
+      world.scene.add(dropShadows)
+    }
+
+    let index = 0
+    sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
+
+    for (const entity of shadowComponentQuery()) {
+      const group = getComponent(entity, GroupComponent)
+      if (!group) continue
+
+      const setDropShadowMatrix = (matrix: Matrix4) => {
+        dropShadows.setMatrixAt(index, matrix)
+        index++
+      }
+
+      const transform = getComponent(entity, TransformComponent)
+
+      const raycaster = new Raycaster()
+      raycaster.layers.disable(ObjectLayers.DropShadowCaster)
+      raycaster.firstHitOnly = true
+      raycaster.set(transform.position, shadowDirection)
+
+      const intersects = raycaster.intersectObjects(sceneObjects)
+      if (!intersects.length || !intersects[0].face) {
+        setDropShadowMatrix(new Matrix4())
```

Comment:
any math objects (Matrix4, Box3, Quaternion etc) that are not going to persist after calculations have been done should be made into scratch variables to minimize garbage collection